### PR TITLE
[WIP][Nek] Update BP1 and add BP3

### PR DESCRIPTION
**Pull request by thilinarmtb**
Also, added a fix so that E-vectors don't need to provide an indices array.

### Summary
- Updated the libCEED reference implementation
- Updated the Fortran interface
- Updated the Nek5000 example

### TODO
- [ ] If the fix is acceptable, fix it in other backends

### CODE REVIEW

**Comment by thilinarmtb** [comment]
@jedbrown Thanks for the reviews !! Have been busy this week, will have a look during the weekend.

**Comment by thilinarmtb** [comment]
@jedbrown : I added a named constant for E-vector identity. Let me know if this looks okay.

File: backends/ref/ceed-ref-restriction.c [hunk 1]
```diff
@@ -26,7 +26,7 @@ static int CeedElemRestrictionApply_Ref(CeedElemRestriction r,
   int ierr;
   const CeedScalar *uu;
   CeedScalar *vv;
-  CeedInt esize = r->nelem*r->elemsize;
+  CeedInt esize = r->nelem*r->elemsize*ncomp;
```
Comment (thilinarmtb) [inline backends/ref/ceed-ref-restriction.c]: @jedbrown : Is this change okay? We should take into account number of
components as well right?

Comment (thilinarmtb) [reply backends/ref/ceed-ref-restriction.c]: Nevermind, I fix this in eb5d317

**Comment by jeremylt** [comment]
It is the way the front end is set up.

`int CeedElemRestrictionApply(CeedElemRestriction r, CeedTransposeMode tmode,
                             CeedInt ncomp, CeedTransposeMode lmode,
                             CeedVector u, CeedVector v, CeedRequest *request) {
  CeedInt m,n;
  int ierr;

  if (tmode == CEED_NOTRANSPOSE) {
    m = r->nelem * r->elemsize * ncomp;
    n = r->ndof * ncomp;
  } else {
    m = r->ndof * ncomp;
    n = r->nelem * r->elemsize * ncomp;
  }
  if (n != u->length)
    return CeedError(r->ceed, 2,
                     "Input vector size %d not compatible with element restriction (%d,%d)",
                     u->length, m, n);
  if (m != v->length)
    return CeedError(r->ceed, 2,
                     "Output vector size %d not compatible with element restriction (%d,%d)",
                     v->length, m, n);
  ierr = r->Apply(r, tmode, ncomp, lmode, u, v, request); CeedChk(ierr);
  return 0;
}`

**Comment by jeremylt** [comment]
@thilinarmtb Once we merge the active-passive update and the new CPU backend, I am going to be running some comparisons to the NEK code in https://github.com/CEED/benchmarks. Will the work in this branch help make that comparison more fair/reasonable?

**Comment by thilinarmtb** [comment]
@jeremylt : We still do a redundant data copy for the Fortran interface here:
https://github.com/CEED/libCEED/blob/master/ceed-fortran.c#L94

I need to fix this; There is alignment problem between Fortran address and the C address.
If you ignore this data copy time, I think the comparison should be fair.

**Comment by jedbrown** [comment]
@thilinarmtb Can we merge this as is (without your planned work to avoid data copies)? You can start a new branch for that after we merge #41.

**Comment by thilinarmtb** [comment]
Hi Jed,

Sure, I am fine merging as it is.

**Comment by jeremylt** [comment]
We should make a conscious choice about names for CEED_ELEMRESTRICT_IDENTITY and CEED_RESTRICTION_IDENTITY before we merge, but everything else looks ready to me.

**Comment by thilinarmtb** [comment]
@jeremylt : Yes, I think they are the same as well. I am fine with either names. So the choice is
yours :)

**Pull request merged**